File: examples/cpp/interpreter.cpp
#include <iostream>
#include <memory>
#include <string>
#include <sstream>
#include <vector>

class IExpression {
public:
    virtual ~IExpression() = default;
    virtual int interpret() const = 0;
};

class NumberExpr : public IExpression {
    int value_;
public:
    NumberExpr(int v) : value_(v) {}
    int interpret() const override { return value_; }
};

class AddExpr : public IExpression {
    std::unique_ptr<IExpression> left_, right_;
public:
    AddExpr(std::unique_ptr<IExpression> l, std::unique_ptr<IExpression> r)
        : left_(std::move(l)), right_(std::move(r)) {}
    int interpret() const override {
        return left_->interpret() + right_->interpret();
    }
};

class SubtractExpr : public IExpression {
    std::unique_ptr<IExpression> left_, right_;
public:
    SubtractExpr(std::unique_ptr<IExpression> l, std::unique_ptr<IExpression> r)
        : left_(std::move(l)), right_(std::move(r)) {}
    int interpret() const override {
        return left_->interpret() - right_->interpret();
    }
};

std::unique_ptr<IExpression> parse(const std::string& expression) {
    std::istringstream iss(expression);
    int num;
    iss >> num;
    std::unique_ptr<IExpression> result = std::make_unique<NumberExpr>(num);

    std::string op;
    while (iss >> op >> num) {
        auto right = std::make_unique<NumberExpr>(num);
        if (op == "+")
            result = std::make_unique<AddExpr>(std::move(result), std::move(right));
        else
            result = std::make_unique<SubtractExpr>(std::move(result), std::move(right));
    }
    return result;
}

int main() {
    std::cout << "=== Interpreter: Arithmetic Expressions ===\n\n";

    std::vector<std::string> expressions = {"3 + 5", "10 - 2 + 4", "100 - 50 - 25"};
    for (const auto& expr : expressions) {
        auto parsed = parse(expr);
        std::cout << "  " << expr << " = " << parsed->interpret() << "\n";
    }
    return 0;
}

File: examples/csharp/Patterns/Interpreter.cs
namespace DesignPatterns.Interpreter
{
    // Simple expression language: "3 + 5 - 2"
    public interface IExpression
    {
        int Interpret();
    }

    public class NumberExpression : IExpression
    {
        private readonly int _value;
        public NumberExpression(int value) => _value = value;
        public int Interpret() => _value;
    }

    public class AddExpression : IExpression
    {
        private readonly IExpression _left, _right;
        public AddExpression(IExpression left, IExpression right) { _left = left; _right = right; }
        public int Interpret() => _left.Interpret() + _right.Interpret();
    }

    public class SubtractExpression : IExpression
    {
        private readonly IExpression _left, _right;
        public SubtractExpression(IExpression left, IExpression right) { _left = left; _right = right; }
        public int Interpret() => _left.Interpret() - _right.Interpret();
    }

    public static class Parser
    {
        public static IExpression Parse(string expression)
        {
            var tokens = expression.Split(' ');
            IExpression result = new NumberExpression(int.Parse(tokens[0]));

            for (int i = 1; i < tokens.Length; i += 2)
            {
                var op = tokens[i];
                var right = new NumberExpression(int.Parse(tokens[i + 1]));
                result = op == "+" ? new AddExpression(result, right)
                                   : new SubtractExpression(result, right);
            }
            return result;
        }
    }

    public static class Demo
    {
        public static void Run()
        {
            var expressions = new[] { "3 + 5", "10 - 2 + 4", "100 - 50 - 25" };
            foreach (var expr in expressions)
            {
                var parsed = Parser.Parse(expr);
                Console.WriteLine($"  {expr} = {parsed.Interpret()}");
            }
        }
    }
}
